### ISSUE TITLE
Theme breaks, if locale is "C"

### ISSUE THREAD

**Issue by dritter** [bhilburn/powerlevel9k#56]
Recently I stumbled upon https://github.com/robbyrussell/oh-my-zsh/pull/4091 which affects powerlevel9k too. I started fixing this problem in my branch [dritter/character_protection](https://github.com/dritter/powerlevel9k/tree/dritter/character_protection), but it is harder than I thought..

The right prompt seems to complicate the problem. Maybe you have some good Ideas to fix that. ;)

Steps to reproduce:
Set `export LANG=C` in your `~/.zshrc`, open a new shell and voilá..

**Comment by dritter** [comment]
Interestingly special characters seem to need to be wrapped in `%{ %}` to avoid unwanted spaces. This is just the case, if there is a right side prompt, which is why I set up a [gist](https://gist.github.com/dritter/ae0f0ebb2926f184838c) of the original agnoster theme with a time-segment on the right side (theme is based on PR 4091 in OMZ; RPROMPT based on powerlevel9k).

**Comment by dritter** [comment]
The pure prompt-string looks quite good:

```
setopt xtrace
+omz_termsupport_precmd:1> [[ '' == true ]]
+omz_termsupport_precmd:5> title '%15<..<%~%<<' '%n@%m: %~'
+title:1> [[ '' == *term* ]]
+title:5> : '%n@%m: %~'
+title:7> [[ xterm-256color == screen* ]]
+title:9> [[ xterm-256color == xterm* ]]
+title:10> print -Pn '\e]2;%n@%m:\ %~\a'
+title:11> print -Pn '\e]1;%15\<..\<%~%\<\<\a'
+omz_termsupport_cwd:3> [[ iTerm.app == Apple_Terminal ]]
+vcs_info:1> emulate -L zsh
+vcs_info:2> setopt extendedglob NO_warn_create_global
+vcs_info:4> [[ -r . ]]
+vcs_info:6> local pat
+vcs_info:7> local -i found retval
+vcs_info:8> local -a enabled disabled dps
+vcs_info:9> local -x usercontext vcs rrn quiltmode LC_MESSAGES
+vcs_info:10> local -ix maxexports
+vcs_info:11> local -ax msgs
+vcs_info:12> local -Ax vcs_comm hook_com backend_misc user_data
+vcs_info:14> LC_MESSAGES=C
+vcs_info:15> [[ -n '' ]]
+vcs_info:20> vcs=-init-
+vcs_info:20> rrn=-all-
+vcs_info:20> quiltmode=addon
+vcs_info:21> usercontext=default
+vcs_info:23> VCS_INFO_hook start-up
+VCS_INFO_hook:5> local hook static func
+VCS_INFO_hook:6> local -x context hook_name
+VCS_INFO_hook:7> local -xi ret
+VCS_INFO_hook:8> local -a hooks tmp
+VCS_INFO_hook:9> local -i debug
+VCS_INFO_hook:11> ret=0
+VCS_INFO_hook:12> hook_name=start-up
+VCS_INFO_hook:13> shift
+VCS_INFO_hook:14> context=:vcs_info:-init-+start-up:default:-all-
+VCS_INFO_hook:15> static=:vcs_info-static_hooks:start-up
+VCS_INFO_hook:17> zstyle -t :vcs_info:-init-+start-up:default:-all- debug
+VCS_INFO_hook:17> debug=0
+VCS_INFO_hook:18> ((  debug  ))
+VCS_INFO_hook:24> zstyle -a :vcs_info-static_hooks:start-up hooks hooks
+VCS_INFO_hook:25> ((  debug  ))
+VCS_INFO_hook:28> zstyle -a :vcs_info:-init-+start-up:default:-all- hooks tmp
+VCS_INFO_hook:29> ((  debug  ))
+VCS_INFO_hook:32> hooks+=( )
+VCS_INFO_hook:33> ((  0 == 0  ))
+VCS_INFO_hook:33> return 0
+vcs_info:24> retval=0
+vcs_info:25> ((  retval == 1  ))
+vcs_info:27> ((  retval == 2  ))
+vcs_info:35> zstyle -a :vcs_info:-init-:default:-all- enable enabled
+vcs_info:36> ((  2 == 0  ))
+vcs_info:38> [[ -n '' ]]
+vcs_info:43> [[ -n '' ]]
+vcs_info:48> zstyle -a :vcs_info:-init-:default:-all- disable-patterns dps
+vcs_info:57> VCS_INFO_maxexports
+VCS_INFO_maxexports:5> setopt localoptions NO_shwordsplit
+VCS_INFO_maxexports:7> zstyle -s :vcs_info:-init-:default:-all- max-exports maxexports
+VCS_INFO_maxexports:7> maxexports=2
+VCS_INFO_maxexports:8> [[ 2 != <-> ]]
+VCS_INFO_maxexports:8> ((  maxexports < 1  ))
+VCS_INFO_maxexports:13> return 0
+vcs_info:59> ((  found = 0  ))
+vcs_info:60> vcs=git
+vcs_info:61> [[ -n '' ]]
+vcs_info:62> ((  1 == 0  ))
+vcs_info:67> vcs_comm=( )
+vcs_info:68> VCS_INFO_get_cmd
+VCS_INFO_get_cmd:4> local cmd
+VCS_INFO_get_cmd:5> zstyle -s :vcs_info:git:default:-all- command cmd
+VCS_INFO_get_cmd:6> vcs_comm[cmd]=git
+vcs_info:69> VCS_INFO_detect_git
+VCS_INFO_detect_git:5> setopt localoptions NO_shwordsplit
+VCS_INFO_detect_git:7> [[ '' == --flavours ]]
+VCS_INFO_detect_git:9> VCS_INFO_check_com git
+VCS_INFO_check_com:5> setopt localoptions NO_shwordsplit
+VCS_INFO_check_com:7> case git (/*)
+VCS_INFO_check_com:7> case git (*)
+VCS_INFO_check_com:12> ((  1  ))
+VCS_INFO_check_com:12> return 0
+VCS_INFO_detect_git:9> git rev-parse --is-inside-work-tree
+VCS_INFO_detect_git:15> return 1
+vcs_info:60> vcs=hg
+vcs_info:61> [[ -n '' ]]
+vcs_info:62> ((  1 == 0  ))
+vcs_info:67> vcs_comm=( )
+vcs_info:68> VCS_INFO_get_cmd
+VCS_INFO_get_cmd:4> local cmd
+VCS_INFO_get_cmd:5> zstyle -s :vcs_info:hg:default:-all- command cmd
+VCS_INFO_get_cmd:6> vcs_comm[cmd]=hg
+vcs_info:69> VCS_INFO_detect_hg
+VCS_INFO_detect_hg:5> setopt localoptions NO_shwordsplit
+VCS_INFO_detect_hg:7> [[ '' == --flavours ]]
+VCS_INFO_detect_hg:9> VCS_INFO_check_com hg
+VCS_INFO_check_com:5> setopt localoptions NO_shwordsplit
+VCS_INFO_check_com:7> case hg (/*)
+VCS_INFO_check_com:7> case hg (*)
+VCS_INFO_check_com:12> ((  1  ))
+VCS_INFO_check_com:12> return 0
+VCS_INFO_detect_hg:10> vcs_comm[detect_need_file]='store data sharedpath'
+VCS_INFO_detect_hg:11> VCS_INFO_bydir_detect .hg
+VCS_INFO_bydir_detect:5> setopt localoptions NO_shwordsplit
+VCS_INFO_bydir_detect:6> local 'dirname=.hg'
+VCS_INFO_bydir_detect:7> local 'basedir=.' realbasedir file
+VCS_INFO_bydir_detect:9> realbasedir=+VCS_INFO_bydir_detect:9> VCS_INFO_realpath .
+VCS_INFO_realpath:5> setopt localoptions NO_shwordsplit chaselinks
+VCS_INFO_realpath:7> cd -q .
+VCS_INFO_realpath:7> pwd
+VCS_INFO_bydir_detect:9> realbasedir=/Users/dr
+VCS_INFO_bydir_detect:10> [[ /Users/dr != / ]]
+VCS_INFO_bydir_detect:11> [[ -r /Users/dr ]]
+VCS_INFO_bydir_detect:12> [[ -n 'store data sharedpath' ]]
+VCS_INFO_bydir_detect:13> [[ -d ./.hg ]]
+VCS_INFO_bydir_detect:22> basedir=./..
+VCS_INFO_bydir_detect:23> realbasedir=+VCS_INFO_bydir_detect:23> VCS_INFO_realpath ./..
+VCS_INFO_realpath:5> setopt localoptions NO_shwordsplit chaselinks
+VCS_INFO_realpath:7> cd -q ./..
+VCS_INFO_realpath:7> pwd
+VCS_INFO_bydir_detect:23> realbasedir=/Users
+VCS_INFO_bydir_detect:10> [[ /Users != / ]]
+VCS_INFO_bydir_detect:11> [[ -r /Users ]]
+VCS_INFO_bydir_detect:12> [[ -n 'store data sharedpath' ]]
+VCS_INFO_bydir_detect:13> [[ -d ./../.hg ]]
+VCS_INFO_bydir_detect:22> basedir=./../..
+VCS_INFO_bydir_detect:23> realbasedir=+VCS_INFO_bydir_detect:23> VCS_INFO_realpath ./../..
+VCS_INFO_realpath:5> setopt localoptions NO_shwordsplit chaselinks
+VCS_INFO_realpath:7> cd -q ./../..
+VCS_INFO_realpath:7> pwd
+VCS_INFO_bydir_detect:23> realbasedir=/
+VCS_INFO_bydir_detect:10> [[ / != / ]]
+VCS_INFO_bydir_detect:26> [[ / == / ]]
+VCS_INFO_bydir_detect:26> return 1
+VCS_INFO_detect_hg:11> return 1
+vcs_info:72> ((  found == 0  ))
+vcs_info:73> vcs=-quilt-
+vcs_info:73> quiltmode=standalone
+vcs_info:74> VCS_INFO_quilt standalone
+VCS_INFO_quilt:1> emulate -L zsh
+VCS_INFO_quilt:2> setopt extendedglob
+VCS_INFO_quilt:3> local 'mode=standalone'
+VCS_INFO_quilt:4> local patches pc tmp qstring root
+VCS_INFO_quilt:5> local -i ret
+VCS_INFO_quilt:6> local -x context
+VCS_INFO_quilt:7> local -a applied unapplied all applied_string unapplied_string quiltcommand
+VCS_INFO_quilt:8> local -Ax hook_com
+VCS_INFO_quilt:10> context=:vcs_info:-quilt-.quilt-standalone:default:-all-
+VCS_INFO_quilt:11> zstyle -t :vcs_info:-quilt-.quilt-standalone:default:-all- use-quilt
+VCS_INFO_quilt:11> return 1
+vcs_info:74> VCS_INFO_set --nvcs
+VCS_INFO_set:5> setopt localoptions noksharrays NO_shwordsplit
+VCS_INFO_set:6> local -i i j
+VCS_INFO_set:8> [[ --nvcs == --nvcs ]]
+VCS_INFO_set:9> [[ '' == -preinit- ]]
+VCS_INFO_set:10> i=0
+VCS_INFO_set:11> typeset -gx 'vcs_info_msg_0_='
+VCS_INFO_set:10> i=1
+VCS_INFO_set:11> typeset -gx 'vcs_info_msg_1_='
+VCS_INFO_set:13> VCS_INFO_nvcsformats
+VCS_INFO_nvcsformats:5> setopt localoptions noksharrays NO_shwordsplit
+VCS_INFO_nvcsformats:6> local c v rr
+VCS_INFO_nvcsformats:8> [[ '' == -preinit- ]]
+VCS_INFO_nvcsformats:13> zstyle -a :vcs_info:-quilt-:default:-all- nvcsformats msgs
+VCS_INFO_nvcsformats:14> ((  0 > maxexports  ))
+VCS_INFO_nvcsformats:15> return 0
+VCS_INFO_set:14> [[ '' != -preinit- ]]
+VCS_INFO_set:14> VCS_INFO_hook no-vcs
+VCS_INFO_hook:5> local hook static func
+VCS_INFO_hook:6> local -x context hook_name
+VCS_INFO_hook:7> local -xi ret
+VCS_INFO_hook:8> local -a hooks tmp
+VCS_INFO_hook:9> local -i debug
+VCS_INFO_hook:11> ret=0
+VCS_INFO_hook:12> hook_name=no-vcs
+VCS_INFO_hook:13> shift
+VCS_INFO_hook:14> context=:vcs_info:-quilt-+no-vcs:default:-all-
+VCS_INFO_hook:15> static=:vcs_info-static_hooks:no-vcs
+VCS_INFO_hook:17> zstyle -t :vcs_info:-quilt-+no-vcs:default:-all- debug
+VCS_INFO_hook:17> debug=0
+VCS_INFO_hook:18> ((  debug  ))
+VCS_INFO_hook:24> zstyle -a :vcs_info-static_hooks:no-vcs hooks hooks
+VCS_INFO_hook:25> ((  debug  ))
+VCS_INFO_hook:28> zstyle -a :vcs_info:-quilt-+no-vcs:default:-all- hooks tmp
+VCS_INFO_hook:29> ((  debug  ))
+VCS_INFO_hook:32> hooks+=( )
+VCS_INFO_hook:33> ((  0 == 0  ))
+VCS_INFO_hook:33> return 0
+VCS_INFO_set:17> ((  0 - 1 < 0  ))
+VCS_INFO_set:17> return 0
+vcs_info:75> return 0
+/opt/local/bin/zsh:2> build_left_prompt
+build_left_prompt:1> [[ 2 == 0 ]]
+build_left_prompt:5> element=dir
+build_left_prompt:6> prompt_dir left
+prompt_dir:1> local 'current_path=%~'
+prompt_dir:2> [[ -n 2 ]]
+prompt_dir:4> current_path='%3(c:.../:)%2c'
+prompt_dir:7> left_prompt_segment prompt_dir blue black '%3(c:.../:)%2c'
+left_prompt_segment:7> local 'BACKGROUND_USER_VARIABLE=POWERLEVEL9K_DIR_BACKGROUND'
+left_prompt_segment:8> local 'BG_COLOR_MODIFIER='
+left_prompt_segment:9> [[ -n '' ]]
+left_prompt_segment:12> local 'FOREGROUND_USER_VARIABLE=POWERLEVEL9K_DIR_FOREGROUND'
+left_prompt_segment:13> local 'FG_COLOR_MODIFIER='
+left_prompt_segment:14> [[ -n '' ]]
+left_prompt_segment:16> local bg fg
+left_prompt_segment:17> [[ -n blue ]]
+left_prompt_segment:17> bg='%K{blue}'
+left_prompt_segment:18> [[ -n black ]]
+left_prompt_segment:18> fg='%F{black}'
+left_prompt_segment:19> [[ NONE != NONE ]]
+left_prompt_segment:24> echo -n '%{%K{blue}%}%{%F{black}%} '
+left_prompt_segment:26> CURRENT_BG=blue
+left_prompt_segment:27> [[ -n '%3(c:.../:)%2c' ]]
+left_prompt_segment:27> echo -n '%3(c:.../:)%2c '
+build_left_prompt:5> element=vcs
+build_left_prompt:6> prompt_vcs left
+prompt_vcs:1> local 'vcs_prompt='
+prompt_vcs:3> [[ -n '' ]]
+build_left_prompt:9> left_prompt_end
+left_prompt_end:1> [[ -n blue ]]
+left_prompt_end:2> echo -n '%{%k%F{blue}%}%{%}'
+left_prompt_end:6> echo -n '%{%f%}'
+left_prompt_end:7> CURRENT_BG=''
+/opt/local/bin/zsh:2> build_right_prompt
+build_right_prompt:1> RETVAL=0
+build_right_prompt:3> [[ 3 == 0 ]]
+build_right_prompt:7> element=aws
+build_right_prompt:8> prompt_aws right
+prompt_aws:1> local 'aws_profile='
+prompt_aws:2> [[ -n '' ]]
+build_right_prompt:7> element=symfony2_version
+build_right_prompt:8> prompt_symfony2_version right
+prompt_symfony2_version:1> [[ -f app/bootstrap.php.cache ]]
+build_right_prompt:7> element=time
+build_right_prompt:8> prompt_time right
+prompt_time:1> local 'time_format=%D{%H:%M:%S}'
+prompt_time:2> [[ -n '' ]]
+prompt_time:6> right_prompt_segment prompt_time white black '%D{%H:%M:%S}'
+right_prompt_segment:2> local 'BACKGROUND_USER_VARIABLE=POWERLEVEL9K_TIME_BACKGROUND'
+right_prompt_segment:3> local 'BG_COLOR_MODIFIER='
+right_prompt_segment:4> [[ -n '' ]]
+right_prompt_segment:7> local 'FOREGROUND_USER_VARIABLE=POWERLEVEL9K_TIME_FOREGROUND'
+right_prompt_segment:8> local 'FG_COLOR_MODIFIER='
+right_prompt_segment:9> [[ -n '' ]]
+right_prompt_segment:11> local bg fg
+right_prompt_segment:12> [[ -n white ]]
+right_prompt_segment:12> bg='%K{white}'
+right_prompt_segment:13> [[ -n black ]]
+right_prompt_segment:13> fg='%F{black}'
+right_prompt_segment:14> echo -n '%{%F{white}%}%{%}%{%f%}%{%K{white}%}%{%F{black}%} '
+right_prompt_segment:15> [[ -n '%D{%H:%M:%S}' ]]
+right_prompt_segment:15> echo -n '%D{%H:%M:%S} %{%f%}'

**Comment by bhilburn** [comment]
Hey @dritter - Do you have a branch where you have tried to integrate @apjanke's changes into our code? Seems like a good place to start. Let me know the branch name and I'll pull it and start debugging, too.

**Comment by apjanke** [comment]
FYI: The stuff about the `%{...%}` is described in the ["Prompt Expansion: Visual Effects"](http://zsh.sourceforge.net/Doc/Release/Prompt-Expansion.html#Visual-effects) section of the `zsh` reference manual. Basically, literal characters which do not change the cursor position (like terminal escape sequences) need to be wrapped in that to let zsh know they're "zero-width" or non-displayable. Except `zsh`'s own prompt `%`-escape sequences like `%F`/`%f`, `%K`/`%k`, or `%B`/`%b`, they're "pre-wrapped", because they happen at the level of `zsh`'s prompt expansion code, so `zsh` is aware of their displayed width. (That's largely why I prefer `%F`/`%f` to other color-control codes; you can skip the `%{...%}` and it ends up more readable IMHO.) Otherwise it'll assume each character uses 1 character cell in the screen's character grid, which could cause miscalculation of where to write characters or place the cursor, especially upon redisplays.

Since you're using `%F` in a lot of places, it looks like a lot of the `%{...%}` additions in your `character_protection` branch are unnecessary.

**Comment by apjanke** [comment]
(Sorry for just popping up in your comments here uninvited, but I got a Notification about this b/c bhilburn at-mentioned me, and I couldn't resist chiming in, since I spent quite a while sorting this stuff out myself recently...)

**Comment by dritter** [comment]
@bhilburn Yes, I created branch `dritter/character_protection` a while ago, but I wanted `dritter/customizable_icons` to be merged first, as this benefits more users. Also, I had some trouble to get it working properly. I mean the characters are not the problem, but aligning the segments correctly is quite hard.

@apjanke No need to apology. I love github for the collaboration and this is the best example. Thanks for the info :) I didn't know that zsh's internal escape sequences are auto-escaped.

As soon as I am through with the `customizable_icons` branch, I continue to work on this one.

**Comment by bhilburn** [comment]
@apjanke - I really appreciate you hopping in! You clearly have a lot of expertise / knowledge about these items, and it's very helpful that you are willing to share some of your experience.

**Comment by bhilburn** [comment]
Fixed by #90
